### PR TITLE
#184347107-search-products

### PR DESCRIPTION
**Pull request by Ntarecedrick**
adding the search route

### What does this PR do?

this PR is a feature to search for a product depending on a user needs

### Description of Task to be completed?

this PR contains codes that will allow a user to be able to search products in our database so he/she can purchase it from out e-commerce website

### Any background context you want to provide?

N/A

### What are the relevant pivotal tracker stories?

https://www.pivotaltracker.com/n/projects/2628461/stories/184347107

### Questions:

# Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] My code is formatted correctly
- [x] My changes doesn't have any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/middlewares/searchMiddleware.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        }
+      } else {
+        next();
+      }
+    } catch (error) {
+      res.status(500).json({ status: 500, message: error });
+    }
+  }
+
+  static async checkDate(req, res, next) {
```
Comment (Eli250) [inline src/middlewares/searchMiddleware.js]: This should be done in validations and use `joi` to validate this.

**Review by NG-Solennel — DISMISSED** [review]

File: src/controllers/productController.js [hunk 2]
```diff
@@ -7,6 +7,7 @@ class ProductController {
       await ProductServices.addItem(req);
       return res.status(200).json({ status: 200, message: "Item added" });
     } catch (error) {
+      console.log(error);
```
Comment (iradukundairenee) [inline src/controllers/productController.js]: remove this console

**Review by iradukundairenee — DISMISSED** [review]

**Review by NG-Solennel — DISMISSED** [review]

File: src/middlewares/badSeach.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/* eslint-disable no-else-return */
```
Comment (Eli250) [inline src/middlewares/badSeach.js]: There is no need for this middleware. Return the response in the search controller

File: src/routes/api/product.route.js [hunk 4]
```diff
@@ -17,17 +19,22 @@ route.post(
...
+route.get("/", extractToken, ProductController.getAllItems);
+
 route.get(
-  "/",
-  extractToken,
-  checkRole(["admin", "seller", "buyer"]),
-  ProductController.getAllItems
+  "/search",
+  validationOfQueries,
+  noResultFound,
```
Comment (Eli250) [inline src/routes/api/product.route.js]: Get rid o this middleware

File: src/utils/findCategory.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { Product, Category } from "../database/models/index";
+
+const displayCategory = async (data) => {
```
Comment (Eli250) [inline src/utils/findCategory.js]: Do this in search services, not in utility.

**Review by Eli250 — CHANGES_REQUESTED** [review]

File: src/middlewares/searchMiddleware.js (hunk 1 above)
Comment (Ntarecedrick) [reply src/middlewares/searchMiddleware.js]: i no longer have this file i created the one of joi

File: src/utils/findCategory.js (hunk 5 above)
Comment (Ntarecedrick) [reply src/utils/findCategory.js]: > Do this in search services, not in utility.

i don't even use this file let me get rid of it

**Review by Eli250 — APPROVED** [review]

**Pull request merged**